firefox-ios/Client/Frontend/Onboarding/Protocols/OnboardingService.swift

```diff
@@ -0,0 +1,293 @@
+        windowUUID: WindowUUID,
+        profile: Profile,
+        themeManager: ThemeManager,
+        delegate: OnboardingServiceDelegate,
+        navigationDelegate: OnboardingNavigationDelegate?,
+        qrCodeNavigationHandler: QRCodeNavigationHandler?
+    ) {
+        self.delegate = delegate
+        self.userDefaults = userDefaults
+        self.windowUUID = windowUUID
+        self.profile = profile
+        self.themeManager = themeManager
+        self.introScreenManager = IntroScreenManager(prefs: profile.prefs)
+        self.navigationDelegate = navigationDelegate
+        self.qrCodeNavigationHandler = qrCodeNavigationHandler
+    }
+
+    func handleAction(
+        _ action: OnboardingActions,
+        from cardName: String,
+        cards: [OnboardingKitCardInfoModel],
+        with activityEventHelper: ActivityEventHelper,
+        completion: @escaping (Result<OnboardingFlowViewModel<OnboardingKitCardInfoModel>.TabAction, Error>) -> Void
+    ) {
+        switch action {
+        case .requestNotifications:
+            handleRequestNotifications(from: cardName, with: activityEventHelper)
+            completion(.success(.none))
+
+        case .forwardOneCard:
+            completion(.success(.advance(numberOfPages: 1)))
+
+        case .forwardTwoCard:
+            completion(.success(.advance(numberOfPages: 2)))
+
+        case .forwardThreeCard:
+            completion(.success(.advance(numberOfPages: 3)))
+
+        case .syncSignIn:
+            handleSyncSignIn(from: cardName, with: activityEventHelper) {
+                completion(.success(.advance(numberOfPages: 1)))
+            }
+
+        case .setDefaultBrowser:
+            handleSetDefaultBrowser(with: activityEventHelper)
+            completion(.success(.none))
+
+        case .openInstructionsPopup:
+            guard let popupViewModel = cards
+                .first(where: { $0.name == cardName })?
+                .instructionsPopup
+            else { return }
+            handleOpenInstructionsPopup(
+                from: OnboardingInstructionsPopupInfoModel(
+                    title: popupViewModel.title,
+                    instructionSteps: popupViewModel.instructionSteps,
+                    buttonTitle: popupViewModel.buttonTitle,
+                    buttonAction: popupViewModel.buttonAction,
+                    a11yIdRoot: popupViewModel.a11yIdRoot
+                )
+            )
+
+        case .readPrivacyPolicy:
+            guard let infoModel = cards
+                .first(where: { $0.name == cardName }),
+                  let url = infoModel.link?.url
+            else { return }
+            handleReadPrivacyPolicy(from: url)
+
+        case .openIosFxSettings:
+            handleOpenIosFxSettings(from: cardName)
+
+        case .endOnboarding:
+            handleEndOnboarding(with: activityEventHelper)
+        }
+    }
+}
+
+private extension OnboardingService {
```

Comment:
I don't think we should use an extension here. Let's add these method to the service directly.